src/content/1.2/code/ocaml/snippet05.ml

```diff
@@ -1 +1,14 @@
-let fact n = List.fold (List.range 1 n) ~init:1 ~f:( * )
+(* OCaml doesn't ship with a lazy sequence range syntax nor a product
+   function; defining our own. *)
+
+let rec range seq n1 n2 =
+  if n2 < n1 then seq
+  else n2 |> pred |> range (fun () -> Seq.Cons (n2, seq)) n1
+let range = range Seq.empty
+
+let rec product result seq = match seq () with
+  | Seq.Nil -> result
+  | Seq.Cons (n, seq) -> product (result * n) seq
+let product = product 1
+
+let fact n = product (range 1 n)
```

Comment:
why not just
```ocaml
let factorial n = Seq.fold_left ( * ) 1 (Seq.init n succ)
```
or
```ocaml
let product = Seq.fold_left ( * ) 1
let factorial n = Seq.init n succ |> product
```
or use `List` instead of `Seq` (at your choice, in case you prefer to be more backwards-compatible) ?
It doesn't really matter that it's "lazy" does it? At least the book doesn't seem to make a point out of it.